src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs

```diff
@@ -0,0 +1,708 @@
+            AssertFloatingPointIncompatible_Fails<sbyte?>();
+            AssertFloatingPointIncompatible_Fails<short?>();
+            AssertFloatingPointIncompatible_Fails<int?>();
+            AssertFloatingPointIncompatible_Fails<long?>();
+            AssertFloatingPointIncompatible_Fails<ushort?>();
+            AssertFloatingPointIncompatible_Fails<uint?>();
+            AssertFloatingPointIncompatible_Fails<ulong?>();
+            AssertFloatingPointIncompatible_Fails<float?>();
+            AssertFloatingPointIncompatible_Fails<decimal?>();
+        }
+
+        private static void AssertFloatingPointIncompatible_Fails<T>()
+        {
+            string[] testCases = new[]
+            {
+                "NaN",
+                "Infinity",
+                "-Infinity",
+            };
+
+            foreach (string test in testCases)
+            {
+                Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<T>(test, s_optionReadFromStrAllowFloatConstants));
+            }
+        }
+
+        [Fact]
+        public static void UnsupportedFormats()
+        {
+            AssertUnsupportedFormatThrows<byte>();
+            AssertUnsupportedFormatThrows<sbyte>();
+            AssertUnsupportedFormatThrows<short>();
+            AssertUnsupportedFormatThrows<int>();
+            AssertUnsupportedFormatThrows<long>();
+            AssertUnsupportedFormatThrows<ushort>();
+            AssertUnsupportedFormatThrows<uint>();
+            AssertUnsupportedFormatThrows<ulong>();
+            AssertUnsupportedFormatThrows<float>();
+            AssertUnsupportedFormatThrows<decimal>();
+            AssertUnsupportedFormatThrows<byte?>();
+            AssertUnsupportedFormatThrows<sbyte?>();
+            AssertUnsupportedFormatThrows<short?>();
+            AssertUnsupportedFormatThrows<int?>();
+            AssertUnsupportedFormatThrows<long?>();
+            AssertUnsupportedFormatThrows<ushort?>();
+            AssertUnsupportedFormatThrows<uint?>();
+            AssertUnsupportedFormatThrows<ulong?>();
+            AssertUnsupportedFormatThrows<float?>();
+            AssertUnsupportedFormatThrows<decimal?>();
+        }
+
+        private static void AssertUnsupportedFormatThrows<T>()
+        {
+            string[] testCases = new[]
+            {
+                " $123.46", // Currency
+                "100.00 %", // Percent
+                 "1234,57", // Fixed point
+                 "00FF", // Hexadecimal
+            };
+
+            foreach (string test in testCases)
+            {
+                Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<T>(test, s_optionReadFromStr));
+            }
+        }
+
+        [Fact]
+        public static void CustomConverterOverridesBuiltInLogic()
+        {
+            var options = new JsonSerializerOptions(s_optionReadAndWriteFromStr)
+            {
+                Converters = { new ConverterForInt32() }
```

Comment:
You should also try with a float converter and attest that `"NaN"` calls into the `JsonConverter.Read` method.